STONEBLD-78 Add initial JVM docs (#109)
redhat-appstudio/book: book/java-rebuilds-service.md

> # JVM Build Service

Reviewer: The file is called `java-rebuilds-service` while the title is `JVM Build Service`. Is one of these more accurate?

Author: JVM is more accurate, I will update the file name.

## Overview

> The JVM Build Service (JBS) is a controller that will rebuild Java and other JVM Language based dependencies from source.

Reviewer: Can you spell out what JVM means here (and potentially in the page title)?

> Java uses a binary distribution model, so the only way to ensure an application is completly built from source is to rebuild all its component libraries from source themselves. Unfortunatly due to the nature of the Java ecosystem this is not an easy process and would normally require a large amount of manual toil.

Reviewer: Do you want to go into why someone might _want_ to go through the complex/toilsome process of rebuilding everything from scratch?

Author: I have added some more context here.

> JBS will attempt to automate this proccess as much as possible, making it much less time consuming to create fully supply chain controlled Java builds.

Reviewer: ```suggestion
JBS automates this process as much as possible, making it much less time consuming to create fully supply chain controlled Java builds.
```

### Dependencies

> JBS depends on Tekton, but is otherwise largely independent of the rest of RHTAP. Given the right configuration it can be deployed outside of RHTAP, which is mainly used for testing and development.

Reviewer: ```suggestion
JBS depends on Tekton, but is otherwise largely independent of the rest of AppStudio. Given the right configuration it can be deployed outside of AppStudio, which is mainly used for testing and development.
```

## Architecture

### Flow Overview

> The JVM Build service can be enabled on a specific namespace by creating a `JBSConfig` object, with `enableRebuilds: true` specified. This will trigger the controller to create a deployment of the local Cache in the namespace. This cache has a number of purposes, but the primary one is to cache artifacts from Maven central to speed up Java builds.

Reviewer: Can you provide a sample object?

Author: I have added it towards the bottom of the document.

> Once it is enabled and the cache has started the JVM Build Service will watch for `PipelineRun` objects with the `JAVA_COMMUNITY_DEPENDENCIES` results. This will be present on end user builds that have dependencies from untrusted upstream sources.

Reviewer: You introduce the concept of untrusted upstream sources here but you do not go into further detail. Can you include/expound upon this term in the Overview section?

Author: I have added a little bit. Do we want both this and the end user docs to be standalone documents? I was thinking we should just make sure all this is explained really well in the end user docs, and then explicitly point out that this doc assumes you have read the other one.

Reviewer: I think as much as reasonable, it makes sense to have each document live on its own. This separation can allow you to be more technical here if desired while not going into as much detail on the user documentation.

If you think linking to the end user documentation is sufficient then I'll leave that up to you!

> When one of these `PipelineRun` objects is observed JBS will extra the dependency list, and attempt to rebuild all the listed artifacts from source.

Reviewer: ```suggestion
When one of these `PipelineRun` objects is observed JBS will extract the dependency list, and attempt to rebuild all the listed artifacts from source.
```

The first stage of this process is to create an `ArtifactBuild` object. This object represents a Maven `group:artifact:version` (GAV) coordinate of the library that needs to be built from source.

Once the `ArtifactBuild` object has been created JBS will then attempt to try and build it. This first step is to attempt to find the relevant source code.

Once the source code location has been discovered a `DependencyBuild` object is created. There are generally less `DependencyBuild` objects than there are `ArtifactBuild`, as multiple artifacts can come from the same build. The controller will then try and build the artifact, first it will run a build discovery pipeline, that attempts to determine possible ways of building the artifact. Once discovery is complete it uses this information to attempt to build the artifact in a trial and error manner.

> If a build is successful the results are stored in a container image and the state is marked as complete, otherwise it is marked as failed and manual effort is required to fix the build.

Reviewer: Are there common types of "manual effort" that may be needed?

Author: Yes, but that is really an end user concern: https://github.com/redhat-appstudio/docs.appstudio.io/blob/main/docs/modules/ROOT/pages/how-to-guides/Secure-your-supply-chain/proc_java_dependencies.adoc#dealing-with-failed-builds

How much overlap do we want to have between these docs?

Reviewer: If you don't want to duplicate information between the documents, can you link to that document from this one?

### Components

JBS consists of the following components:

**Controller**

The controller is a Kubernetes controller written in go that orchestrates all aspects of the rebuild process. This will kick off all build related pipelines, and manage the state of the relevant Kubernetes objects.

**Cache**

The cache is a Quarkus Java application that caches artifacts from upstream repositories such as maven central. It also performs the following additional functions:

- Handles container image based dependencies that have been rebuilt
- Injects tracking metadata into class files to detect contaminated builds
- Looks up source and build information from the build recipe database

All dependency rebuilds are configured to only get their artifacts from the cache.

**Build Request Processor**

This is a multipurpose Quarkus CLI based app that performs quite a few different functions depending on the parameters it is invoked with. This is packaged up into an image, and then run in different parts of various pipelines by the operator.

> In essence any custom logic that is needed in a pipeline goes in here, and then the operator will invoke it with the correct arguments. This is a single multi-tool rather than individual functions as experience has shown this is a lot easier to manage. It's functions include:

Reviewer: What is a "single multi-tool"?

I don't think you need to include "as experience has shown" here.

Author: I have tidied it up. Its just a single binary with lots of different functions.

- Analysing a build and looking up its build information in the recipe database

> - Preprocessing a build to fix common problems, such as removing problematic plugins

Reviewer: Are there examples of problematic plugins? Does this affect any functionality on the end application built? Are these problematic plugins customizable?

Author: One would be the nexus deploy plugin, which disables the standard deployment plugin and instead assumes you are deploying to a nexus server. They don't affect the functionality of the final application.
